source/frontend/src/features/admin/financial-codes/add/AddFinancialCodeContainer.test.tsx

```diff
@@ -0,0 +1,102 @@
+import { createMemoryHistory } from 'history';
+import { mockFinancialCode } from 'mocks';
+import { Api_FinancialCode } from 'models/api/FinancialCode';
+import { act, createAxiosError, render, RenderOptions, screen } from 'utils/test-utils';
+
+import AddFinancialCodeContainer, { IAddFinancialCodeFormProps } from './AddFinancialCodeContainer';
+
+const mockApi = {
+  error: undefined,
+  response: undefined,
+  execute: jest.fn(),
+  loading: false,
+};
+
+jest.mock('../hooks/useFinancialCodeRepository', () => ({
+  useFinancialCodeRepository: () => {
+    return {
+      addFinancialCode: mockApi,
+    };
+  },
+}));
+
+const history = createMemoryHistory();
+
+let viewProps: IAddFinancialCodeFormProps | undefined;
+
+const TestView: React.FC<IAddFinancialCodeFormProps> = props => {
+  viewProps = props;
+  return <span>Content Rendered</span>;
+};
+
+describe('AddFinancialCode container', () => {
+  const setup = (renderOptions: RenderOptions = {}) => {
+    const utils = render(<AddFinancialCodeContainer View={TestView} />, {
+      ...renderOptions,
+      history,
+    });
+    return { ...utils };
+  };
+
+  beforeEach(() => {
+    viewProps = undefined;
+    jest.resetAllMocks();
+  });
+
+  it('renders the underlying form', () => {
+    const { getByText } = setup();
+    expect(getByText(/Content Rendered/)).toBeVisible();
+  });
+
+  it('displays error message for duplicate codes', async () => {
+    const { getByText } = setup();
+
+    await act(async () => {
+      const error409 = createAxiosError(409, 'Duplicate code found');
+      viewProps?.onError(error409);
+    });
+
+    expect(getByText(/Cannot create duplicate financial code/)).toBeVisible();
+  });
+
+  it('displays a toast for generic server errors', async () => {
+    setup();
+
+    await act(async () => {
+      const error500 = createAxiosError(500);
+      viewProps?.onError(error500);
```

Comment:
having direct access to the props seems nice, looks like it is easier to write concise tests.

Reply:
Yeah, I like this new pattern. It allows to "ignore" which button was clicked in the underlying form as that is no the concern of the container. Happy to demo it this Friday as I still see new PRs not using this pattern

Reply:
If you see any new containers not using dependency injection please block them from being merged. (changes required)